miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -269,6 +328,39 @@ impl<'p> IntoIterator for &'p SparseMerklePath {
     }
 }
 
+/// An iterator over nodes known by a [SparseMerklePath]. See
+/// [`SparseMerklePath::authenticated_nodes()`].
+pub struct InnerNodeIterator<'p> {
+    path: &'p SparseMerklePath,
+    index: NodeIndex,
+    value: Word,
+}
+
+impl Iterator for InnerNodeIterator<'_> {
+    type Item = InnerNodeInfo;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index.is_root() {
+            return None;
+        }
+
+        let index_depth = NonZero::new(self.index.depth()).expect("non-root depth cannot be 0");
+        let path_node = self.path.at_depth(index_depth).unwrap();
+
+        let is_right = self.index.is_value_odd();
+        let (left, right) = if is_right {
+            (path_node, self.value)
+        } else {
+            (self.value, path_node)
+        };
+
+        self.value = Rpo256::merge(&[left, right]);
```

Comment:
I think we can simplify this by using `NodeIndex::build_node()`. I'll do this in a separate commit.